<a href="https://colab.research.google.com/github/Amit198181/JBVNL/blob/main/MDI_0_AUG_100.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files
import os

uploaded = files.upload()

for fn in uploaded.keys():
    print('User uploaded file "{name}" with length {length} bytes'.format(
        name=fn, length=len(uploaded[fn])))

Saving LTIS_APR_AUG.xlsx to LTIS_APR_AUG.xlsx
User uploaded file "LTIS_APR_AUG.xlsx" with length 2230734 bytes


Load data: Load the data from the uploaded Excel file "LTIS_APR_AUG.xlsx" into a pandas DataFrame.
Filter for august data: Filter the DataFrame to include only the data for the month of August.
Filter for maximum demand of zero: Filter the August data to include only consumers with a maximum demand of zero.
Filter for consumption of 100 units or more: From the filtered data, identify consumers with 100 units or more consumption.
Identify unique consumers: Extract the unique consumer identifiers from the final filtered data.
Display results: Display the list of consumers who meet all the criteria.
Finish task: Summarize the findings and present the list of consumers.

# Task
Find out the consumers having the maximum demand of zero in the month of August, with 100 units and more consumption of electricity from the file "LTIS_APR_AUG.xlsx".

## Load data

### Subtask:
Load the data from the uploaded Excel file "LTIS_APR_AUG.xlsx" into a pandas DataFrame.


**Reasoning**:
Import pandas and load the Excel file into a DataFrame, then display the head and info.



In [2]:
import pandas as pd

df = pd.read_excel('LTIS_APR_AUG.xlsx')
display(df.head())
display(df.info())

ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

  CONSUMER_TYPE  ... LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
0         URBAN  ...                    0.0                      0.0   
1         RURAL  ...                    0.0                      0.0   
2         RURAL  ...                    0.0                      0.0   
3         RURAL  ...                    0.0                      0.0   
4         RURAL  ...                    0.0                      0.0   

  BILLED_UNITS_HTS_202507 MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
0                    95.0           12.0                     0.0   
1                    30.0            5.0                     0.0   
2                    61.0            6.0                     0.0   
3                    55.0            4.0                     0.0   
4                    25.0            5.0                     0.0   

   SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
0                      0.0                    275.0            18.0   
1                      0.0                     70.0             4.0   
2                      0.0                    120.0             5.0   
3                      0.0                    150.0             6.0   
4                      0.0                    102.0             6.0   

   LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
0                     0.0                        0  
1                     0.0                        0  
2                     0.0                        0  
3                     0.0                        0  
4                     0.0                        0  

[5 rows x 30 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17742 entries, 0 to 17741
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ZONE                     15715 non-null  object 
 1   CIRCLE                   15715 non-null  object 
 2   DIVISION                 15715 non-null  object 
 3   CONSNO                   15715 non-null  object 
 4   ACCOUNT_NO               17742 non-null  int64  
 5   CONSUMER_STATUS          15715 non-null  object 
 6   CONSUMER_NAME            15714 non-null  object 
 7   TARIFF_CODE              15715 non-null  object 
 8   CONTRACTED_LOAD          15715 non-null  float64
 9   CONSUMER_TYPE            15715 non-null  object 
 10  METER_READER_NAME        15703 non-null  object 
 11  MF_KWH                   15715 non-null  float64
 12  PREPAID_FLAG             15715 non-null  object 
 13  MTR_CONSUMP_TYPE         15715 non-null  object 
 14  BILLED_UNITS_HTS_20250

None

## Filter for august data

### Subtask:
Filter the DataFrame to include only the data for the month of August.


**Reasoning**:
Filter the DataFrame to include only the data for the month of August by selecting rows where the `BILLED_UNITS_HTS_202508` column is not null.



In [3]:
df_august = df[df['BILLED_UNITS_HTS_202508'].notnull()].copy()
display(df_august.head())
display(df_august.info())

ZONE  CIRCLE     DIVISION   CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
0  RANCHI  RANCHI  RANCHI WEST  RT27482  11220279599            LIVE   
1  RANCHI  RANCHI  RANCHI WEST  RT28138  11220279600            LIVE   
2  RANCHI  RANCHI  RANCHI WEST  RT27405  11220279612            LIVE   
3  RANCHI  RANCHI  RANCHI WEST  RT28556  11220279628            LIVE   
4  RANCHI  RANCHI  RANCHI WEST  RT28602  11220279636            LIVE   

                               CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
0                M/S SABOO SONS PRATIK SABOO        LTIS             40.0   
1                               SUVESH KUMAR        LTIS              8.0   
2                               TAMIM ANSARI        LTIS              6.0   
3  SARITA SRIVASTAVA, W/O - S K SRIVASTAVA,         LTIS             10.0   
4                         SHWETA FABRICATION        LTIS             10.0   

  CONSUMER_TYPE  ... LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
0         URBAN  ...                    0.0                      0.0   
1         RURAL  ...                    0.0                      0.0   
2         RURAL  ...                    0.0                      0.0   
3         RURAL  ...                    0.0                      0.0   
4         RURAL  ...                    0.0                      0.0   

  BILLED_UNITS_HTS_202507 MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
0                    95.0           12.0                     0.0   
1                    30.0            5.0                     0.0   
2                    61.0            6.0                     0.0   
3                    55.0            4.0                     0.0   
4                    25.0            5.0                     0.0   

   SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
0                      0.0                    275.0            18.0   
1                      0.0                     70.0             4.0   
2                      0.0                    120.0             5.0   
3                      0.0                    150.0             6.0   
4                      0.0                    102.0             6.0   

   LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
0                     0.0                        0  
1                     0.0                        0  
2                     0.0                        0  
3                     0.0                        0  
4                     0.0                        0  

[5 rows x 30 columns]

<class 'pandas.core.frame.DataFrame'>
Index: 15715 entries, 0 to 17741
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ZONE                     15715 non-null  object 
 1   CIRCLE                   15715 non-null  object 
 2   DIVISION                 15715 non-null  object 
 3   CONSNO                   15715 non-null  object 
 4   ACCOUNT_NO               15715 non-null  int64  
 5   CONSUMER_STATUS          15715 non-null  object 
 6   CONSUMER_NAME            15714 non-null  object 
 7   TARIFF_CODE              15715 non-null  object 
 8   CONTRACTED_LOAD          15715 non-null  float64
 9   CONSUMER_TYPE            15715 non-null  object 
 10  METER_READER_NAME        15703 non-null  object 
 11  MF_KWH                   15715 non-null  float64
 12  PREPAID_FLAG             15715 non-null  object 
 13  MTR_CONSUMP_TYPE         15715 non-null  object 
 14  BILLED_UNITS_HTS_202504  14

None

## Filter for maximum demand of zero

### Subtask:
Filter the August data to include only consumers with a maximum demand of zero.


**Reasoning**:
Filter the `df_august` DataFrame to select rows where the 'MDI_HTS_202508' column is equal to 0 and display the head and information of the resulting DataFrame.



In [4]:
df_august_zero_demand = df_august[df_august['MDI_HTS_202508'] == 0].copy()
display(df_august_zero_demand.head())
display(df_august_zero_demand.info())

ZONE  CIRCLE     DIVISION  CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
46  RANCHI  RANCHI  RANCHI WEST     GB8  11220717002            LIVE   
67  RANCHI  RANCHI  RANCHI EAST   M1820  11220768009            LIVE   
68  RANCHI  RANCHI  RANCHI EAST   M1821  11220768017            LIVE   
79  RANCHI  RANCHI  RANCHI EAST  RP3993  11220784596            LIVE   
89  RANCHI  RANCHI  RANCHI EAST   S1126  11220806457            LIVE   

                       CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
46    ASSTT.ENGINEER D.W.R.S SUB-DIV        LTIS             7.02   
67    ASSISTANT MECH.ENGINEER P.H.ME        LTIS            20.00   
68  ASSISTANT MECH ENG.PH.MECH.SUB          LTIS            20.00   
79                     PRATIMA ORAON        LTIS            20.00   
89         EXECUTIVE ENGINEER,PHED          LTIS             7.00   

   CONSUMER_TYPE  ... LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
46          GOVT  ...                    0.0                      0.0   
67        OTHERS  ...                    0.0                      0.0   
68        OTHERS  ...                    0.0                      0.0   
79        OTHERS  ...                    NaN                      NaN   
89        OTHERS  ...                    0.0                      0.0   

   BILLED_UNITS_HTS_202507 MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
46                   404.0           0.00                     0.0   
67                  2938.0           0.00                     0.0   
68                    93.0           0.00                     0.0   
79                     0.0           0.01                     0.0   
89                  1000.0           0.00                     0.0   

    SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
46                      0.0                   1213.0             0.0   
67                      0.0                   2693.0             0.0   
68                      0.0                     85.0             0.0   
79                      0.0                      0.0             0.0   
89                      0.0                   1109.0             0.0   

    LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
46                     0.0                        0  
67                     0.0                        0  
68                     0.0                        0  
79                     0.0                        0  
89                     0.0                        0  

[5 rows x 30 columns]

<class 'pandas.core.frame.DataFrame'>
Index: 1342 entries, 46 to 17733
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ZONE                     1342 non-null   object 
 1   CIRCLE                   1342 non-null   object 
 2   DIVISION                 1342 non-null   object 
 3   CONSNO                   1342 non-null   object 
 4   ACCOUNT_NO               1342 non-null   int64  
 5   CONSUMER_STATUS          1342 non-null   object 
 6   CONSUMER_NAME            1342 non-null   object 
 7   TARIFF_CODE              1342 non-null   object 
 8   CONTRACTED_LOAD          1342 non-null   float64
 9   CONSUMER_TYPE            1342 non-null   object 
 10  METER_READER_NAME        1342 non-null   object 
 11  MF_KWH                   1342 non-null   float64
 12  PREPAID_FLAG             1342 non-null   object 
 13  MTR_CONSUMP_TYPE         1342 non-null   object 
 14  BILLED_UNITS_HTS_202504  12

None

## Filter for consumption of 100 units or more

### Subtask:
From the filtered data, identify consumers with 100 units or more consumption.


**Reasoning**:
Filter the DataFrame to include only consumers with 100 units or more consumption.



In [5]:
df_august_high_consumption = df_august_zero_demand[df_august_zero_demand['BILLED_UNITS_HTS_202508'] >= 100].copy()
display(df_august_high_consumption.head())
display(df_august_high_consumption.info())

ZONE  CIRCLE     DIVISION  CONSNO   ACCOUNT_NO CONSUMER_STATUS  \
46   RANCHI  RANCHI  RANCHI WEST     GB8  11220717002            LIVE   
67   RANCHI  RANCHI  RANCHI EAST   M1820  11220768009            LIVE   
89   RANCHI  RANCHI  RANCHI EAST   S1126  11220806457            LIVE   
120  RANCHI  RANCHI  RANCHI EAST  TT7830  11220850561            LIVE   
205  RANCHI  RANCHI  RANCHI EAST  S18783  11220955749            LIVE   

                      CONSUMER_NAME TARIFF_CODE  CONTRACTED_LOAD  \
46   ASSTT.ENGINEER D.W.R.S SUB-DIV        LTIS             7.02   
67   ASSISTANT MECH.ENGINEER P.H.ME        LTIS            20.00   
89        EXECUTIVE ENGINEER,PHED          LTIS             7.00   
120     BAJRANG PRESTRESSAD PRODUCT        LTIS            88.00   
205         Praveen Kumar Dumriya          LTIS            10.00   

    CONSUMER_TYPE  ... LOAD_FACTOR_HTS_202506  SUPPLY_HOURS_HTS_202506  \
46           GOVT  ...                    0.0                      0.0   
67         OTHERS  ...                    0.0                      0.0   
89         OTHERS  ...                    0.0                      0.0   
120        OTHERS  ...                    0.0                      0.0   
205       PRIVATE  ...                    0.0                      0.0   

    BILLED_UNITS_HTS_202507 MDI_HTS_202507  LOAD_FACTOR_HTS_202507  \
46                    404.0            0.0                     0.0   
67                   2938.0            0.0                     0.0   
89                   1000.0            0.0                     0.0   
120                  5936.0            0.0                     0.0   
205                  3475.0            0.0                     0.0   

     SUPPLY_HOURS_HTS_202507  BILLED_UNITS_HTS_202508  MDI_HTS_202508  \
46                       0.0                   1213.0             0.0   
67                       0.0                   2693.0             0.0   
89                       0.0                   1109.0             0.0   
120                      0.0                   3909.0             0.0   
205                      0.0                   3277.0             0.0   

     LOAD_FACTOR_HTS_202508  SUPPLY_HOURS_HTS_202508  
46                      0.0                        0  
67                      0.0                        0  
89                      0.0                        0  
120                     0.0                        0  
205                     0.0                        0  

[5 rows x 30 columns]

<class 'pandas.core.frame.DataFrame'>
Index: 990 entries, 46 to 17733
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ZONE                     990 non-null    object 
 1   CIRCLE                   990 non-null    object 
 2   DIVISION                 990 non-null    object 
 3   CONSNO                   990 non-null    object 
 4   ACCOUNT_NO               990 non-null    int64  
 5   CONSUMER_STATUS          990 non-null    object 
 6   CONSUMER_NAME            990 non-null    object 
 7   TARIFF_CODE              990 non-null    object 
 8   CONTRACTED_LOAD          990 non-null    float64
 9   CONSUMER_TYPE            990 non-null    object 
 10  METER_READER_NAME        990 non-null    object 
 11  MF_KWH                   990 non-null    float64
 12  PREPAID_FLAG             990 non-null    object 
 13  MTR_CONSUMP_TYPE         990 non-null    object 
 14  BILLED_UNITS_HTS_202504  918

None

**Reasoning**:
Extract the unique consumer identifiers from the final filtered data and display the list of unique consumers.



In [6]:
unique_consumers = df_august_high_consumption['CONSNO'].unique()
print("Unique consumers with zero maximum demand and 100 or more units consumption in August:")
print(unique_consumers)

Unique consumers with zero maximum demand and 100 or more units consumption in August:
['GB8' 'M1820' 'S1126' 'TT7830' 'S18783' 'OM10033' 'OM241G' 'OM9710'
 'KOL11025' 'CHU15722' 'CHY23239' 'L374' 'L375' 'LK37' 'BIT3663' 'LK106LT'
 'LK108LT' 'LK127LT' 'LK131LT' 'NS7359' 'LK140LT' 'LK144LT' 'LK152LT'
 'LK159LT' 'LK176LT' 'LK178LT' 'LK187LT' 'UB3474' 'HE4331' 'HW2896'
 'HW2897' 'AGOVT02' 'HN14827' 'HOK5477' 'HECS1995' 'HECN1419' 'HECS2168'
 'HECSP1563' 'PM99' 'RD3946' 'RRD455A' 'PM4963' 'PM403' 'PM11413'
 'PM11616' 'PM12023' 'PM12455' 'PM32918' 'PM33579' 'PM38968' 'pm39897'
 'rd31764' 'rrd35490' 'PM39906' '532' '537' '1330' '1394' '22508' '23123'
 '65026' '20T14302' 'GUI15660' 'PLI28134' 'PLI29456' '68421' '68544'
 '68804' 'GRI12899' 'GRI33567' 'GRI33605' 'GUI17984' 'GRI65222' '497'
 '502' '6920' '6928' '6946' '9158' '9336' '13068' '13461' '985' '40955'
 '41971' 'LO48603' 'LADLT51701' 'LADLT51903' 'LADLT51906' 'LADLT51908'
 '4859' '4860' '16132' 'LTIS0025' 'ITT9877' 'RT14780' 'RT4766' 'C

## Identify unique consumers

### Subtask:
Extract the unique consumer identifiers from the final filtered data.


**Reasoning**:
Extract the unique consumer identifiers from the 'CONSNO' column of the filtered data and store them in a variable.



In [7]:
unique_consumers = df_august_high_consumption['CONSNO'].unique()
print("Unique consumers with zero maximum demand and 100 or more units consumption in August:")
print(unique_consumers)

Unique consumers with zero maximum demand and 100 or more units consumption in August:
['GB8' 'M1820' 'S1126' 'TT7830' 'S18783' 'OM10033' 'OM241G' 'OM9710'
 'KOL11025' 'CHU15722' 'CHY23239' 'L374' 'L375' 'LK37' 'BIT3663' 'LK106LT'
 'LK108LT' 'LK127LT' 'LK131LT' 'NS7359' 'LK140LT' 'LK144LT' 'LK152LT'
 'LK159LT' 'LK176LT' 'LK178LT' 'LK187LT' 'UB3474' 'HE4331' 'HW2896'
 'HW2897' 'AGOVT02' 'HN14827' 'HOK5477' 'HECS1995' 'HECN1419' 'HECS2168'
 'HECSP1563' 'PM99' 'RD3946' 'RRD455A' 'PM4963' 'PM403' 'PM11413'
 'PM11616' 'PM12023' 'PM12455' 'PM32918' 'PM33579' 'PM38968' 'pm39897'
 'rd31764' 'rrd35490' 'PM39906' '532' '537' '1330' '1394' '22508' '23123'
 '65026' '20T14302' 'GUI15660' 'PLI28134' 'PLI29456' '68421' '68544'
 '68804' 'GRI12899' 'GRI33567' 'GRI33605' 'GUI17984' 'GRI65222' '497'
 '502' '6920' '6928' '6946' '9158' '9336' '13068' '13461' '985' '40955'
 '41971' 'LO48603' 'LADLT51701' 'LADLT51903' 'LADLT51906' 'LADLT51908'
 '4859' '4860' '16132' 'LTIS0025' 'ITT9877' 'RT14780' 'RT4766' 'C

## Display results

### Subtask:
Display the list of consumers who meet all the criteria.


**Reasoning**:
Display the list of unique consumers who meet all the criteria.



In [8]:
print("List of unique consumers with zero maximum demand and 100 or more units consumption in August:")
for consumer in unique_consumers:
    print(consumer)

List of unique consumers with zero maximum demand and 100 or more units consumption in August:
GB8
M1820
S1126
TT7830
S18783
OM10033
OM241G
OM9710
KOL11025
CHU15722
CHY23239
L374
L375
LK37
BIT3663
LK106LT
LK108LT
LK127LT
LK131LT
NS7359
LK140LT
LK144LT
LK152LT
LK159LT
LK176LT
LK178LT
LK187LT
UB3474
HE4331
HW2896
HW2897
AGOVT02
HN14827
HOK5477
HECS1995
HECN1419
HECS2168
HECSP1563
PM99
RD3946
RRD455A
PM4963
PM403
PM11413
PM11616
PM12023
PM12455
PM32918
PM33579
PM38968
pm39897
rd31764
rrd35490
PM39906
532
537
1330
1394
22508
23123
65026
20T14302
GUI15660
PLI28134
PLI29456
68421
68544
68804
GRI12899
GRI33567
GRI33605
GUI17984
GRI65222
497
502
6920
6928
6946
9158
9336
13068
13461
985
40955
41971
LO48603
LADLT51701
LADLT51903
LADLT51906
LADLT51908
4859
4860
16132
LTIS0025
ITT9877
RT14780
RT4766
CL1314
CL1406
CL1415
CLT1420
BD3855
BD3857
BD4315
MD5577
MND2797
MNLT41927
BULT12251
B32639
MD18818
BU12037
BPUR25004
SNDR4408
43189
6238
3713
23288
BPUR25005
KHL0002
566
2456
KH6537
2426
15570
6035
KR

## Summary:

### Data Analysis Key Findings

*   A total of 15,715 records had non-null data for August 2025.
*   Out of the August records, 1,342 consumers had a maximum demand of zero.
*   From the consumers with zero maximum demand in August, 990 had a billed consumption of 100 units or more.
*   The unique consumers identified as having zero maximum demand and 100 or more units of consumption in August were extracted.

### Insights or Next Steps

*   Investigate the characteristics and consumption patterns of the 990 consumers identified to understand why they have zero maximum demand despite significant consumption.
*   Evaluate if these consumers are correctly categorized or if there are data collection or measurement issues related to maximum demand for this group.


In [9]:
df_august_high_consumption.to_excel("MDI_zero_100UNITS.xlsx", index=False)
print("Data downloaded as MDI_zero_100UNITS.xlsx")

Data downloaded as MDI_zero_100UNITS.xlsx
